Drivers: ADC: add PSoC4 SAR ADC driver

Deepika-aerlync opened the pull request:
Add support for Infineon PSoC4 ADC Driver.

Deepika-aerlync committed:
dts: arm: infineon: Add PSoC4 ADC updation device tree file

- Add PSoC4100TP SoC device tree include for ADC.
- Provide the basic structure for ADC channel configuration.

Signed-off-by: Deepika R <deepika@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

Deepika-aerlync committed:
dts: bindings: adc: Add PSoC4 ADC yaml file

- Add device tree bindings for PSoC4 SAR ADC peripheral.
- The binding define the configuration properties
  for PSoC4's Successive Approximation ADC capabilities.

Signed-off-by: Deepika R <deepika@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

Deepika-aerlync committed:
drivers: adc: add PSoC4 ADC driver

Implement ADC driver for PSoC4 with:
- Single-ended and differential input modes.
- SARMUX and AMUXBUS routing support
- Configurable 8, 10, and 12-bit resolution.
- Internal 1.2V BGR and VDDA, VDDA/2, EXTERNAL voltage references.
- Hardware oversampling support.
- Device Tree configuration for all parameters.

Signed-off-by: Deepika R <deepika@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

Deepika-aerlync committed:
samples: drivers: adc: Add ADC overlay

- Add a board-specific devicetree overlay for ADC peripheral testing.
- Enable ADC driver validation.

Signed-off-by: Deepika R <deepika@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

jsbatch commented on drivers/adc/adc_infineon_psoc4_sar.c:
```diff
@@ -0,0 +1,917 @@
+		en_hsiom_sel_t hsiom_sel =
+			(data->channel_cfg[ch].vplus_routing == PSOC4_ADC_ROUTE_AMUXBUS_A)
+				? (en_hsiom_sel_t)6
+				: (en_hsiom_sel_t)7;
```
Is there a HAL define or constant that can be use instead of these magic numbers?  If not, use a descriptive local constant.

jsbatch commented on drivers/adc/adc_infineon_psoc4_sar.c:
```diff
@@ -0,0 +1,917 @@
+		ticks = 3;
+	}
+	if (ticks > 1023) {
+		ticks = 1023;
```
There's a lot of undocumented magic numbers here.  Please use constants, and comment why these values were chosen or where they came from.

jsbatch commented on drivers/adc/adc_infineon_psoc4_sar.c:
```diff
@@ -0,0 +1,917 @@
+
+	uint8_t sample_time_idx = 0xFF;
+
+	for (int i = 0; i < 4; i++) {
```
Same as above.

Deepika-aerlync replied:
Changed.

jsbatch commented on drivers/adc/adc_infineon_psoc4_sar.c:
```diff
@@ -0,0 +1,917 @@
+		data->resolution[ch] = sequence->resolution;
+		const struct psoc4_adc_channel_cfg *ch_cfg = &data->channel_cfg[ch];
+
+		uint32_t sw_mask0 = 0;
```
sw_mask is set, but never really used for anything useful.  I only see it being used to conditionally log an error on line 599.  Either use it for something or remove it from this function, but as it stands now it's just adding a bunch of unneeded complexity to the start_read function.

JarmouniA reviewed (commented):
Add appropriate tests (in Tests/...) support for the driver.

Deepika-aerlync committed:
tests: drivers: adc: Add ADC overlay for test

- Add a board-specific devicetree overlay to enable ADC tests.
- Enable functional testing of the ADC driver.

Signed-off-by: Deepika R <deepika@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

Deepika-aerlync commented on dts/bindings/adc/infineon,psoc4-sar-adc.yaml:
```diff
@@ -0,0 +1,83 @@
+    ADC channel configuration
+
+  properties:
+    zephyr,gpio-port:
```
Removed.

sreeramIfx commented on drivers/adc/adc_infineon_psoc4_sar.c:
```diff
@@ -0,0 +1,992 @@
+
+LOG_MODULE_REGISTER(psoc4_adc, CONFIG_ADC_LOG_LEVEL);
+
+#define PSOC4_ADC_NUM_CHANNELS    CY_SAR_SEQ_NUM_CHANNELS
```
Can we use CY_SAR_SEQ_NUM_CHANNELS directly?

sreeramIfx commented on drivers/adc/adc_infineon_psoc4_sar.c:
```diff
@@ -0,0 +1,992 @@
+	uint32_t channels_mask;
+	uint8_t resolution[PSOC4_ADC_NUM_CHANNELS];
+	struct psoc4_adc_channel_cfg channel_cfg[PSOC4_ADC_NUM_CHANNELS];
+	uint32_t sample_times[4];
```
```suggestion
	uint32_t sample_times[PSOC4_ADC_NUM_SAMPLE_TIMES];
```

Deepika-aerlync replied:
Changed.

laneb-infineon commented on samples/drivers/adc/adc_dt/boards/cy8cproto_041tp.overlay:
```diff
@@ -0,0 +1,101 @@
+		reg = <0>;
+		zephyr,gain = "ADC_GAIN_1";
+		zephyr,reference = "ADC_REF_INTERNAL";
+		zephyr,input-positive = <0>;
```
Instead of using numbers here, we should use macros to correspond to the input-positive and input-negative selections that a User places in their overlay files.

Deepika-aerlync committed:
dt-bindings: adc: add Infineon PSoC4 SAR ADC bindings

- Add a device tree binding header for the Infineon PSoC4 SAR ADC.
  Header provides descriptive macros for:
- SARMUX hardware addresses (0-7) for dedicated port 2 pins.
- GPIO pin mapping macros (Port 0 to Port 6) to simplify ADC channel
  configuration in devicetree overlays.

Signed-off-by: Deepika R <deepika@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

Deepika-aerlync committed:
boards: infineon: cy8cproto_041tp: Configure ADC peripheral clock

- Enable and configure a peripheral clock divider for the SAR ADC.
- Assigns peri_clk_div4 to the ADC resource.

Signed-off-by: Deepika R <deepika@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

Deepika-aerlync committed:
drivers: clock_control: Add support for ADC peripheral clock

Implement peripheral clock configuration support for the SAR ADC

- Update the peripheral clock initialization logic for ADC.

Signed-off-by: Deepika R <deepika@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

laneb-infineon commented on drivers/adc/adc_infineon_psoc4_sar.c:
```diff
@@ -0,0 +1,1097 @@
+		routing = PSOC4_ADC_ROUTE_SARMUX;
+	} else {
+		/* AMUXBUS routing */
+		routing = (pin % 2 == 0) ? PSOC4_ADC_ROUTE_AMUXBUS_A : PSOC4_ADC_ROUTE_AMUXBUS_B;
```
After further testing, discussion, and review, AMUXBUS implementation will not be necessary at this time to introduce the ADC driver to the Zephyr upstream. A bug involving high impedance is present when using AMUXBUS pins, but should not block the ADC introduction PR. I have created a GitHub issue to track a future addition of adding AMUXBUS support. Related GitHub issue: https://github.com/zephyrproject-rtos/zephyr/issues/103374

Instead, this thread can be resolved once AMUXBUS routing has been removed from this implementation. The driver should support SARMUX and SARBUS (CTB OpAmp outputs) routing, but not AMUXBUS. This will limit the pins that a User can select, but will ensure that the selected pins behave accurately with precise voltage readings.
